### ISSUE TITLE
torch.nonzero(t, as_tuple=...) does not work with the JIT because the as_tuple signatures are not exposed properly

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#45499]
```
def foo(t):
  return torch.nonzero(t, as_tuple=False)
: RuntimeError: Arguments for call are not valid.
```

This is because of the Python arg parsing done to handle the "as_tuple" kwarg. Note that tracing works fine.

**Comment by eellison** [comment]
currently it's hard to support this op because the return type depends on the value of `as_tuple`. I think this would not be that hard to support by adding `Literal[True]` and `Literal[False]` and then making the argument rule that we only accept constants for literal arguments. Integration with native_functions.yaml would be a little trickier, or we could just register it as a prim op & separately in python as currently.

**Comment by mruberry** [comment]
> currently it's hard to support this op because the return type depends on the value of `as_tuple`. I think this would not be that hard to support by adding `Literal[True]` and `Literal[False]` and then making the argument rule that we only accept constants for literal arguments. Integration with native_functions.yaml would be a little trickier, or we could just register it as a prim op & separately in python as currently.

Or maybe we should remove the kwarg and deprecate functions that return different types of objects based on their input values.

**Comment by eellison** [comment]
That also sounds good to me! I don't know whether nonzero was on the path to deprecation, or what would be the eager concerns around deprecating it outside of the JIT. I know that numpy has quite a large number of functions that have an output type dependent on the input type.

The mechanism to return a different type dependent on input type is implemented for python, just not for builtins : https://github.com/pytorch/pytorch/blob/bb19a554291aca857e81e9acdf49aae6745f3159/torch/functional.py#L782

**Comment by ngimel** [comment]
@eellison what did you do for `unique`? It also returns either tensor or tuple of tensors depending on the value of kwargs (and btw `unique` has the same behavior in numpy)

**Comment by ppwwyyxx** [comment]
Same issue as https://github.com/pytorch/pytorch/issues/38718

**Comment by eellison** [comment]
@ngimel it uses the dispatching method we have implemented in  
python https://github.com/pytorch/pytorch/blob/master/torch/functional.py#L803. Since `torch.nonzero` already dispatches to two difference aten ops (`nonzero`, and `nonzero_numpy`), maybe we could move nonzero C++ pybind binding to pytorch and support it without a lot of trouble ?

**Comment by rafale77** [comment]
Just ran into this as well. Quite annoying. Looking forward to a fix.

**Comment by eellison** [comment]
I can give guidance on how this might be fixed but I don't know if I have bandwidth to fix this. Adding back to triage.

**Comment by gmagogsfm** [comment]
> I can give guidance on how this might be fixed but I don't know if I have bandwidth to fix this. Adding back to triage.

Could you write up a little about how to fix it in details? Thanks!

**Comment by eellison** [comment]
To do this "the right way" I would :

Add Literal Types - `Literal[True]`, `Literal[False]`
```
@torch.jit.overload
def unique(tensor, as_tuple: Literal[True]): ...

@torch.jit.overload
def unique(tensor, as_tuple: Literal[False]): ...

def unique(tensor, as_tuple: bool):
    if as_tuple:
        return torch._C.unique_tuple(tensor)
   else:
        return torch._C.unique(tensor)
```

But this might be more trouble than it's worth. We might consider throwing an error, exposing the tuple/non-tuple signatures, and asking the user to either directly invoke the unique_tuple vs unique overloads (those may or may not exist rn, i'm not sure).

**Comment by nlgranger** [comment]
FYI, https://github.com/pytorch/pytorch/commit/649e683255c152236a5db8ad86cdfbeb1b4611e1 (folllowing https://github.com/pytorch/pytorch/issues/45499) fixes the type annotations of nonzero.

**Comment by rgommers** [comment]
> But this might be more trouble than it's worth. 

It does seem worth doing - this has already come up for `unique`, `nonzero`, and various `torch.linalg` functions.

**Comment by ppwwyyxx** [comment]
Any updates? It has been a year since #38718 is opened.

**Comment by gmagogsfm** [comment]
Thanks @eellison  for proposing a potential solution. Indeed it looks more trouble (introducing Literal type) than it is worth. Besides, since the output tuple type depends on rank of input tensor, it doesn't seem we can handle the `as_tuple=True` case easily in IR emission either.

What if we model the output as `Any` and asks user, who might have better knowledge about specific type of output tuple to refine the result?

**Comment by mruberry** [comment]
Note that the Python Array API specifies nonzero return a tuple by default, although we may be able to argue it should return a tensor, @ngimel, @rgommers

**Comment by rgommers** [comment]
From https://github.com/data-apis/array-api/pull/23:

_The reason apparently was that it's easier for advanced indexing (you can then simply do `x[*tuple_idx]`, with the non-tuple form it'd be more annoying)._
...
_The tuple form for nonzero allows boolean array indices to be equivalent to nonzero (except for scalar booleans), so `x[boolean_array]` == `x[nonzero(boolean_array)]` (if there are multiple indices, the `nonzero`s are unpacked in the tuple)._

**Comment by ngimel** [comment]
Without tuple, you can do `x[mask.nonzero().unbind(-1)]`, I don't think it's more annoying.

**Comment by rgommers** [comment]
True, `unbind` comes in pretty handy here. That's PyTorch-specific though, and the comments above are about a portable solution (NumPy, TF, Dask, CuPy, MXNet).

**Comment by rgommers** [comment]
I investigated the state some more, a bit of history in NumPy on `nonzero` returning a tuple added in https://github.com/data-apis/array-api/pull/23#issuecomment-858715317. I also searched the NumPy mailing list and issue trackers of NumPy, CuPy, JAX, TensorFlow and Dask, and asked a few maintainers of other libraries. There were no real complaints/issues about `nonzero` returning a tuple.

There were a number of issues related to `nonzero` output shape depending on values in the input and that being a problem for preallocating/JIT-ing. But that's unrelated to the tuple-return thing.

In practice, there's not much of a difference one way or the other - it's just that PyTorch made a different choice than other libraries for return type early on. It's not going to change in the array API or in NumPy. So for that aspect, I think the choice is between PyTorch remaining different, or doing a hard BC-breaking change. There's something to say for the former here.

**Comment by MatthieuToulemont** [comment]
Using torch.where works quite well for me and is scriptable
